File: ui/state/timecode.py
"""Time and timecode helpers for the Streamlit UI adapter."""

from __future__ import annotations


def format_timecode(timestamp_seconds: float) -> str:
    """Format elapsed seconds as an HH:MM:SS.mmm timecode string."""
    total_milliseconds = max(int(round(timestamp_seconds * 1000)), 0)
    hours, remainder = divmod(total_milliseconds, 3_600_000)
    minutes, remainder = divmod(remainder, 60_000)
    seconds, milliseconds = divmod(remainder, 1_000)
    return f"{hours:02d}:{minutes:02d}:{seconds:02d}.{milliseconds:03d}"


def format_time_seconds_for_frame(*, frame_index: int, fps: float) -> float:
    """Convert a frame index into seconds for the current source FPS."""
    if fps <= 0:
        return 0.0
    return max(frame_index, 0) / fps

File: frontend/src/utils/timecode.ts
export function formatTimecode(seconds: number): string {
  const boundedSeconds = Number.isFinite(seconds) ? Math.max(seconds, 0) : 0
  const totalMilliseconds = Math.round(boundedSeconds * 1000)
  const hours = Math.floor(totalMilliseconds / 3_600_000)
  const minutes = Math.floor((totalMilliseconds % 3_600_000) / 60_000)
  const secs = Math.floor((totalMilliseconds % 60_000) / 1000)
  const milliseconds = totalMilliseconds % 1000

  return [
    hours.toString().padStart(2, '0'),
    minutes.toString().padStart(2, '0'),
    secs.toString().padStart(2, '0'),
  ].join(':') + `.${milliseconds.toString().padStart(3, '0')}`
}

export function clampFrameIndex(frameIndex: number, frameCount: number): number {
  if (frameCount <= 0) {
    return 0
  }

  return Math.max(0, Math.min(Math.round(frameIndex), frameCount - 1))
}

export function frameIndexToSeconds(frameIndex: number, fps: number): number {
  if (fps <= 0) {
    return 0
  }

  return clampFrameIndex(frameIndex, Number.MAX_SAFE_INTEGER) / fps
}

export function secondsToFrameIndex(seconds: number, fps: number, frameCount: number): number {
  if (fps <= 0) {
    return 0
  }

  return clampFrameIndex(Math.round(seconds * fps), frameCount)
}
